crates/astria-bridge-withdrawer/src/bridge_withdrawer/submitter/signer/frost_signer.rs

```diff
@@ -0,0 +1,312 @@
+}
+
+impl FrostSigner {
+    pub(crate) async fn sign(&self, tx: TransactionBody) -> eyre::Result<Transaction> {
+        // part 1: gather commitments from participants
+        let round_1_results = self.execute_round_1().await;
+        ensure!(
+            round_1_results.responses.len() >= self.min_signers,
+            "not enough part 1 responses received; want at least {}, got {}",
+            self.min_signers,
+            round_1_results.responses.len()
+        );
+
+        // part 2: gather signature shares from participants
+        let tx_bytes = tx.to_raw().encode_to_vec();
+        let sig_shares = self
+            .execute_round_2(round_1_results.responses, tx_bytes.clone())
+            .await;
+        ensure!(
+            sig_shares.len() >= self.min_signers,
+            "not enough part 2 signature shares received; want at least {}, got {}",
+            self.min_signers,
+            sig_shares.len()
+        );
+
+        // finally, aggregate and create signature
+        let signing_package = frost_ed25519::SigningPackage::new(
+            round_1_results.signing_package_commitments,
+            &tx_bytes,
+        );
+        let signature =
+            frost_ed25519::aggregate(&signing_package, &sig_shares, &self.public_key_package)
+                .wrap_err("failed to aggregate signature shares")?;
+
+        let raw_transaction = astria_core::generated::astria::protocol::transaction::v1::Transaction {
+                body: Some(pbjson_types::Any {
+                    type_url: astria_core::generated::astria::protocol::transaction::v1::TransactionBody::type_url(),
+                    value: tx_bytes.into(),
+                }),
+                signature: signature
+                    .serialize()
+                    .wrap_err("failed to serialize signature")?
+                    .into(),
+                public_key: self.public_key_package
+                    .verifying_key()
+                    .serialize()
+                    .wrap_err("failed to serialize verifying key")?
+                    .into(),
+            };
+        let transaction = Transaction::try_from_raw(raw_transaction)
+            .wrap_err("failed to convert raw transaction to transaction")?;
+
+        Ok(transaction)
+    }
+
+    pub(crate) fn address(&self) -> &Address {
+        &self.address
+    }
+}
+
+struct Round1Results {
+    responses: Vec<Round1Response>,
+    signing_package_commitments: BTreeMap<Identifier, round1::SigningCommitments>,
+}
+
+struct Round1Response {
+    id: Identifier,
+    commitment: axum::body::Bytes,
+    request_identifier: u32,
+}
+
+impl FrostSigner {
+    async fn execute_round_1(&self) -> Round1Results {
+        let stream = futures::stream::FuturesUnordered::new();
+        for (id, client) in &self.participant_clients {
+            let mut client = client.clone();
+            stream.push(async move {
```

Comment:
Put this async block into a freestanding function (+ instrumentation and `err(Display`) to emit an error if something goes wrong).